syncd/syncd_pfc_watchdog.cpp

```diff
@@ -0,0 +1,260 @@
+
+PfcWatchdog& PfcWatchdog::getInstance(void)
+{
+    static PfcWatchdog wd;
+
+    return wd;
+}
+
+sai_object_id_t PfcWatchdog::queueIdToPortId(
+        _In_ sai_object_id_t queueId)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr =
+    {
+        .id = SAI_QUEUE_ATTR_PORT,
+        .value =
+        {
+            .oid = queueId,
+        }
+    };
+
+    sai_status_t status = sai_metadata_sai_queue_api->get_queue_attribute(queueId, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get port Id of queue 0x%lx: %d", queueId, status);
+        return SAI_NULL_OBJECT_ID;
+    }
+
+    return attr.value.oid;
+}
+
+void PfcWatchdog::collectCounters(
+        _In_ swss::Table &countersTable)
+{
+    SWSS_LOG_ENTER();
+
+    std::lock_guard<std::mutex> lock(g_mutex);
+
+    // Collect stats for every registered queue
+    for (const auto &kv: m_counterIdsMap)
+    {
+        const auto &queueVid = kv.first;
+        const auto &queueId = kv.second->queueId;
+        const auto &portId = kv.second->portId;
+        const auto &portCounterIds = kv.second->portCounterIds;
+        const auto &queueCounterIds = kv.second->queueCounterIds;
+
+        std::vector<uint64_t> portStats(portCounterIds.size());
+        std::vector<uint64_t> queueStats(queueCounterIds.size());
+
+        // Get port stats for queue
+        sai_status_t status = sai_metadata_sai_port_api->get_port_stats(
+                portId,
+                static_cast<uint32_t>(portCounterIds.size()),
+                portCounterIds.data(),
+                portStats.data());
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to get stats of port 0x%lx: %d", portId, status);
+            continue;
+        }
+
+        // Get queue stats
+        status = sai_metadata_sai_queue_api->get_queue_stats(
+                queueId,
+                static_cast<uint32_t>(queueCounterIds.size()),
+                queueCounterIds.data(),
+                queueStats.data());
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to get stats of queue 0x%lx: %d", queueVid, status);
+            continue;
+        }
+
+        // Push all counter values to a single vector
+        std::vector<swss::FieldValueTuple> values;
+
+        for (size_t i = 0; i != portCounterIds.size(); i++)
+        {
+            const std::string &counterName = sai_serialize_port_stat(portCounterIds[i]);
+            values.emplace_back(counterName, std::to_string(portStats[i]));
+        }
+
+        for (size_t i = 0; i != queueCounterIds.size(); i++)
+        {
+            const std::string &counterName = sai_serialize_queue_stat(queueCounterIds[i]);
+            values.emplace_back(counterName, std::to_string(queueStats[i]));
+        }
+
+        // Write counters to DB
+        std::string queueVidStr = sai_serialize_object_id(queueVid);
+
+        countersTable.set(queueVidStr, values, "");
```

Comment:
>queueVidStr [](start = 26, length = 11)

your counter table is always index by queue vid, I think it is probably easier for upper layer management?